optax/_src/transform.py

```diff
@@ -594,6 +594,91 @@ def update_fn(updates, state, params=None):
   return base.GradientTransformation(init_fn, update_fn)
 
 
+class ScaleByAdanState(NamedTuple):
+  count: chex.Array
+  mu: base.Updates
+  nu: base.Updates
+  delta: base.Updates
+  grad_tm1: base.Updates
+
+
+def scale_by_adan(
+    b1: float = 0.98,
+    b2: float = 0.92,
+    b3: float = 0.99,
+    eps: float = 1e-8,
+    eps_root: float = 0.0,
+) -> base.GradientTransformation:
+  """Rescale updates according to the Adan algorithm.
+
+  References:
+    [Xie et al, 2022](https://arxiv.org/abs/2208.06677)
+
+  Args:
+    b1: Decay rate for the exponentially weighted average of gradients.
+    b2: Decay rate for the exponentially weighted average of difference of
+      gradients.
+    b3: Decay rate for the exponentially weighted average of the squared term.
+    eps: term added to the denominator to improve numerical stability.
+    eps_root: Term added to the denominator inside the square-root to improve
+      numerical stability when backpropagating gradients through the rescaling.
+
+  Returns:
+    An (init_fn, update_fn) tuple.
+  """
+  def init_fn(params):
+    mu = otu.tree_zeros_like(params)  # 1st moment
+    nu = otu.tree_zeros_like(params)  # 2nd moment
+    delta = otu.tree_zeros_like(params)  # EWA of difference of gradients
+    grad_tm1 = otu.tree_zeros_like(params)  # previous gradient
+    return ScaleByAdanState(
+        count=jnp.zeros((), jnp.int32),
+        mu=mu,
+        nu=nu,
+        delta=delta,
+        grad_tm1=grad_tm1,
+    )
+
+  def update_fn(updates, state, params=None):
```

Comment:
could you add a couple of docstrings explaining what mathematical operations are done on each block? that would be helpful to map the mathematical description of the algorithm to the implementation here

Reply:
@fabianp Do you mean putting comments inside `update_fn`, or extending the docstring of `scale_by_adan`?

Reply:
I've edited `update_fn` to make it more closely reflect Algorithm 1 of https://arxiv.org/pdf/2208.06677v4#page=6, and added a docstring to reflect this.